### PR TITLE
Responder component

### LINKED ISSUE #65: Implement responder component

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes #65 

# Test Plan:
CLI tested

### CODE REVIEW

File: common/trinary/trit_array.c [hunk 1]
```diff
@@ -128,7 +128,7 @@ size_t int8_to_flex_trit_array(flex_trit_t *const to_trit_array,
                                size_t const to_len, trit_t const *const trits,
                                size_t const len, size_t const num_trits) {
   // Bounds checking
-  if (num_trits > len || num_trits > to_len) {
+  if (num_trits > len) {
```
Comment (thibault-martinez) [inline common/trinary/trit_array.c:131]: @ifullgaz Could you have a look at this please, I feel like this check is wrong ?

File: common/network/hash_request.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#define __COMMON_NETWORK_HASH_REQUEST_H__
+
+#include "common/network/neighbor.h"
+#include "common/trinary/trit_array.h"
+
+typedef struct {
+  neighbor_t *neighbor;
+  trit_array_p hash;
+} hash_request_t;
+
```
Comment (th0br0) [inline common/network/hash_request.h]: let's call this `transaction_request_t`?

File: common/network/components/responder.c [hunk 3]
```diff
@@ -6,21 +6,104 @@
...
 #include "common/network/components/responder.h"
+#include "ciri/core.h"
+#include "ciri/node.h"
+#include "common/storage/sql/defs.h"
+#include "common/storage/storage.h"
 #include "utils/logger_helper.h"
 
 #define RESPONDER_COMPONENT_LOGGER_ID "responder_component"
 
+static bool get_transaction_to_reply(responder_state_t *const state,
```
Comment (th0br0) [inline common/network/components/responder.c]: `get_transaction_for_request`?

File: common/network/components/responder.c [hunk 4]
```diff
@@ -6,21 +6,104 @@
...
+                                     hash_request_t *const request,
+                                     iota_transaction_t *const tx) {
+  if (state == NULL) {
```
Comment (th0br0) [inline common/network/components/responder.c]: I think we can just use `||` ?

File: common/network/components/responder.c [hunk 5]
```diff
@@ -6,21 +6,104 @@
...
+    return false;
+  }
+  if (request == NULL) {
+    return false;
+  }
+  if (request->neighbor == NULL) {
```
Comment (th0br0) [inline common/network/components/responder.c:26]: what if the neighbour was removed in between the request being added to the list & being responded to

File: common/network/neighbor.h [hunk 6]
```diff
@@ -21,7 +21,7 @@ typedef struct {
   unsigned int nbr_new_tx;
   unsigned int nbr_invalid_tx;
   unsigned int nbr_sent_tx;
-  unsigned int random_transaction_requests;
+  unsigned int random_tx_req;
```
Comment (th0br0) [inline common/network/neighbor.h]: `nbr_random_tx_req`?

File: common/network/iota_packet.c [hunk 7]
```diff
@@ -22,3 +22,19 @@ void iota_packet_build(iota_packet_t* const packet, size_t const length,
...
+                                 iota_transaction_t const tx) {
+  flex_trit_t tx_flex_trits[FLEX_TRIT_SIZE_8019];
+  trit_t tx_trits[NUM_TRITS_SERIALIZED_TRANSACTION];
+
+  if (packet == NULL || tx == NULL) {
+    return;
+  }
+
+  transaction_serialize_on_flex_trits(tx, tx_flex_trits);
+  flex_trit_array_to_int8(tx_trits, NUM_TRITS_SERIALIZED_TRANSACTION,
```
Comment (th0br0) [inline common/network/iota_packet.c]: :thinking: we need an efficient `flex_trit_array_to_bytes` method I think

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/network/components/responder.c (hunk 4 above)
Comment (thibault-martinez) [reply common/network/components/responder.c]: I was getting it ready for https://github.com/iotaledger/entangled/issues/129 when they'll be different error codes

File: common/network/iota_packet.c (hunk 7 above)
Comment (thibault-martinez) [reply common/network/iota_packet.c]: Agreed

File: common/network/components/responder.c (hunk 5 above)
Comment (thibault-martinez) [reply common/network/components/responder.c:26]: Right, I think we need to copy that neighbor (at least the endpoint part) and let the sending fail

File: common/network/iota_packet.c (hunk 7 above)
Comment (th0br0) [reply common/network/iota_packet.c]: @thibault-martinez could you add a tracker  '`// Tracked via #165` here so we don't forget about it?

Comment (thibault-martinez) [reply common/network/iota_packet.c]: Done

**Review by th0br0 — APPROVED** [review]

**Pull request merged**